engine/graphics/src/graphics_vertices.cpp

```diff
@@ -0,0 +1,510 @@
+//
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#include "graphics.h"
+
+#include <dlib/math.h>
+
+namespace dmGraphics
+{
+	static const float g_SourceDataDefaultVectorEmpty[4]  = {};
+    static const float g_SourceDataDefaultVectorOneAsW[4] = {0.0, 0.0, 0.0, 1.0};
+    static const float g_SourceDataDefaultVector[4]       = {1.0, 1.0, 1.0, 1.0};
+    static const float g_SourceDataDefaultMatrix[16]      = {
+        1.0, 0.0, 0.0, 0.0,
+        0.0, 1.0, 0.0, 0.0,
+        0.0, 0.0, 1.0, 0.0,
+        0.0, 0.0, 0.0, 1.0
+    };
+
+    struct UnpackAttributeData
+    {
+        VertexAttribute::VectorType m_VectorType;
+        VertexAttribute::DataType   m_DataType;
+        const uint8_t*              m_ValuePtr;
+        uint32_t                    m_ElementCount;
+        uint8_t                     m_IsMatrix : 1;
+    };
+
+    void GetAttributeValues(const dmGraphics::VertexAttribute& attribute, const uint8_t** data_ptr, uint32_t* data_size)
+    {
+        *data_ptr  = attribute.m_Values.m_BinaryValues.m_Data;
+        *data_size = attribute.m_Values.m_BinaryValues.m_Count;
+    }
+
+    uint8_t* WriteVertexAttributeFromFloat(uint8_t* value_write_ptr, float value, dmGraphics::VertexAttribute::DataType data_type)
+    {
+        switch (data_type)
+        {
+            case dmGraphics::VertexAttribute::TYPE_BYTE:
+                *((int8_t*) value_write_ptr) = (int8_t) dmMath::Clamp(value, -128.0f, 127.0f);
+                return value_write_ptr + 1;
+            case dmGraphics::VertexAttribute::TYPE_UNSIGNED_BYTE:
+                *value_write_ptr = (uint8_t) dmMath::Clamp(value, 0.0f, 255.0f);
+                return value_write_ptr + 1;
+            case dmGraphics::VertexAttribute::TYPE_SHORT:
+                *((int16_t*) value_write_ptr) = (int16_t) dmMath::Clamp(value, -32768.0f, 32767.0f);
+                return value_write_ptr + 2;
+            case dmGraphics::VertexAttribute::TYPE_UNSIGNED_SHORT:
+                *((uint16_t*) value_write_ptr) = (uint16_t) dmMath::Clamp(value, 0.0f, 65535.0f);
+                return value_write_ptr + 2;
+            case dmGraphics::VertexAttribute::TYPE_INT:
+                *((int32_t*) value_write_ptr) = (int32_t) dmMath::Clamp(value, -2147483648.0f, 2147483647.0f);
+                return value_write_ptr + 4;
+            case dmGraphics::VertexAttribute::TYPE_UNSIGNED_INT:
+                *((uint32_t*) value_write_ptr) = (uint32_t) dmMath::Clamp(value, 0.0f, 4294967295.0f);
+                return value_write_ptr + 4;
+            case dmGraphics::VertexAttribute::TYPE_FLOAT:
+                *((float*) value_write_ptr) = value;
+                return value_write_ptr + 4;
```

Comment:
Future thing: is there float16 support that we want to use as well?

Reply:
Yes, I think there is half_float support for attributes. Could be a future improvement!